File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-2.py
class Solution:
    def lowestCommonAncestor(self, root: TreeNode, p: TreeNode, q: TreeNode) -> TreeNode:
        def dfs(root, p, q):
            if not root:
                return None 
            
            if root.val == p.val or root.val == q.val:
                return root 
                
            left = dfs(root.left, p, q)
            right = dfs(root.right, p, q)
            if left and right:
                return root
            return left or right
        return dfs(root, p, q)

File: Data Structures & Algorithms/lowest-common-ancestor-of-a-binary-tree-iii/submission-2.java
/*
// Definition for a Node.
class Node {
    public int val;
    public Node left;
    public Node right;
    public Node parent;
};
*/

class Solution {
    public Node lowestCommonAncestor(Node p, Node q) {
        return dfs(p, q);
        
    }

    public int depth(Node node) {
        int count = 0;

        while (node != null) {
            count++;
            node = node.parent;
        }

        return count;
    }

    public Node dfs(Node p, Node q){
        
        if (p == null && q == null){
            return null;
        }
        
        if (p == q){
            return p;
        }
        int depth_q = depth(q);
        int depth_p = depth(p);
        
        if (depth_p > depth_q){
            return dfs(p.parent, q);
        }
        else if (depth_p < depth_q){
            return dfs(p, q.parent);
        }
        else{
            return dfs(p.parent, q.parent);
        }
    }

    
}